python/ql/src/semmle/python/regex.qll

```diff
@@ -121,8 +121,57 @@ deprecated string mode_from_mode_object(Value obj) {
 abstract class RegexString extends Expr {
   RegexString() { (this instanceof Bytes or this instanceof Unicode) }
 
+  override string toString() {
+    result = this.(Bytes).getText()
+    or
+    result = this.(Unicode).getText()
+  }
+
+  /** result is true for those start chars that actually mark a start of a char set. */
```

Comment:
Another problem is on line 154, where we recursively invoke `result = char_set_start(p2).booleanNot()`. Rewriting this to not use booleans would mean rewriting this line as `not char_set_start(p2)`, but then we have introduced negative recursion.

Reply:
Clearly I forgot to write proper comments here. This is following the same pattern as `predicate escapingChar(int pos) { this.escaping(pos) = true }` and `private boolean escaping(int pos)` to avoid negative recursion. The pattern is used again in scanning through char sets to find character ranges. Both of these instances are a bit more elaborate than `escaping` and therefor the pattern is likely harder to detect without documentation.

Reply:
I have added more elaborate comments to this effect now.